test/src/main/java/org/apache/accumulo/test/GCRunIT.java

```diff
@@ -0,0 +1,242 @@
+
+  @Override
+  protected Duration defaultTimeout() {
+    return Duration.ofSeconds(30);
+  }
+
+  @BeforeAll
+  public static void setup() throws Exception {
+    SharedMiniClusterBase.startMiniCluster();
+  }
+
+  @AfterAll
+  public static void teardown() {
+    SharedMiniClusterBase.stopMiniCluster();
+  }
+
+  /**
+   * Simple run through GCRun candidate to validate that other tests change behaviour when error
+   * injected.
+   */
+  @Test
+  public void goPath() throws Exception {
+    final String[] names = getUniqueNames(2);
+    final String table1 = names[0];
+    final String clone1 = names[1];
+
+    fillMetadataEntries(table1, clone1);
+
+    try (AccumuloClient client = Accumulo.newClient().from(getClientProps()).build()) {
+
+      var context = getCluster().getServerContext();
+
+      scanReferences(new GCRun(Ample.DataLevel.ROOT, context));
+      scanReferences(new GCRun(Ample.DataLevel.METADATA, context));
+      scanReferences(new GCRun(Ample.DataLevel.USER, context));
+
+      client.tableOperations().delete(clone1);
+    }
+  }
+
+  /**
+   * Force deletion of dir entry from the metadata table - expect GCRun to throw exception scanning
+   * candidate
+   */
+  @Test
+  public void forceMissingDirTest() throws Exception {
+
+    final String[] names = getUniqueNames(2);
+    final String table1 = names[0];
+    final String clone1 = names[1];
+
+    fillMetadataEntries(table1, clone1);
+    try (AccumuloClient client = Accumulo.newClient().from(getClientProps()).build()) {
+
+      client.securityOperations().grantTablePermission(getAdminPrincipal(), MetadataTable.NAME,
+          TablePermission.WRITE);
+
+      String cloneId = client.tableOperations().tableIdMap().get(clone1);
+
+      // pick last row that should always exist.
+      final Mutation m = new Mutation(new Text(cloneId + "<"));
+      final Text colf = new Text(MetadataSchema.TabletsSection.ServerColumnFamily.NAME);
+      final Text colq = new Text(DIRECTORY_QUAL);
+      m.putDelete(colf, colq, new ColumnVisibility());
+
+      try (BatchWriter bw = client.createBatchWriter(MetadataTable.NAME,
+          new BatchWriterConfig().setMaxMemory(Math.max(m.estimatedMemoryUsed(), 1024))
+              .setMaxWriteThreads(1).setTimeout(5_000, TimeUnit.MILLISECONDS))) {
+        log.info("forcing delete of srv:dir with mutation {}", m.prettyPrint());
+        bw.addMutation(m);
+      }
+
+      var context = getCluster().getServerContext();
+
+      scanReferences(new GCRun(Ample.DataLevel.ROOT, context));
+      scanReferences(new GCRun(Ample.DataLevel.METADATA, context));
+      // "missing srv:dir prevents full reference scan for user table
+      assertThrows(IllegalStateException.class,
+          () -> scanReferences(new GCRun(Ample.DataLevel.USER, context)));
+
+      client.tableOperations().delete(clone1);
+    }
+  }
+
+  /**
+   * This is a place holder - deleting the prev row causes the scan to fail before validation check.
```

Comment:
One way to test this would be to make make the code in GCRun that converts a TabletMetadata object to a Reference object into a function.  The could write unit test for that function.  However its not easy to create a tablet metadata object.  In the elasticity branch there is code that makes it easy to create TabletMetadata objects for unit testing via a builder.

Reply:
The entire building of the references in getReferences() in GCRun (starting at line 162) seems that it could benefit from refactoring to improve testing.  However, for a bug-fix release it may be better to keeps changes to a minimum.  Maybe refactoring and additional testing would be best as a separate task in either 3,1 or elasticity?